spec.html

```diff
@@ -11059,6 +11060,44 @@ <h1>Examples of Automatic Semicolon Insertion</h1>
       <p>In the circumstance that an assignment statement must begin with a left parenthesis, it is a good idea for the programmer to provide an explicit semicolon at the end of the preceding statement rather than to rely on automatic semicolon insertion.</p>
     </emu-clause>
   </emu-clause>
+  <emu-clause id="sec-hazards-of-automatic-semicolon-insertion">
+    <h1>Hazards of Automatic Semicolon Insertion</h1>
+    <em>This section is non-normative.</em>
+    <p>ECMAScript programs can be written in a style with very few semicolons by relying on automatic semicolon insertion. As described above, semicolons are not inserted at every newline, and automatic semicolon insertion can depend on multiple tokens across line terminators.</p>
+
+    <p>As new syntactic features are added to ECMAScript, additional grammar productions could be added that cause lines relying on automatic semicolon insertion preceding them to change grammar productions when parsed. In order to prevent cases where automatic semicolon insertion could change by adding source text outside of a grammar productions existing text boundaries, explicit semicolon use is recommended.</p>
+
+    <p>The term &ldquo;<dfn>automatic semicolon insertion hazard</dfn>&rdquo; refers, to a place where a semicolon may or may not be inserted depending on preceding source text according to the rules above. These places are considered hazards if changes outside of an existing source text grammar production could change the grammar production of the existing source text depending on automatic semicolon insertion. The rest of this section describes a number of automatic semicolon insertion hazards in this version of ECMAScript.</p>
+    <emu-clause id="sec-asi-hazards-in-statement-lists">
+      <h1>Automatic Semicolon Insertion Hazards in Statement Lists</h1>
+      <p>In a |StatementList|, many |StatementListItem|s end in semicolons, which may be omitted using automatic semicolon insertion. As a consequence of the rules above, at the end of a line ending an expression, a semicolon is required if the following line begins with any of the following:
```

Comment:
A semi-colon is not required in those scenarios, `void` works as well.

Reply:
that would change the completion value of the statement.

Reply:
TIL the completion value is a thing.

Reading the spec, it only matters at the REPL or when using `eval` though, not in code loaded from files (where coding styles matter and where you'd have to rely on semi-columns to avoid the problems listed below). `void` doesn't turn an abrupt completion into a normal one.

Reply:
It would matter in the upcoming "do expressions" proposal.

Reply:
Sometimes people use eval to load files! It's important to get semantics right here in lots of ways (e.g., we have to do function hoisting out of blocks in sloppy mode that escapes eval).

Reply:
@bakkot 
> Initialized fields, uninitialized fields, static fields, computed property names, async methods, generator methods, and accessors have been on the table or a part of the language with the current syntax for several years, a time period which covers most of the discussions to which which I'm referring.

In fact, this was considered in the original design of ES6 class bodies.  The assumption at the time was that the class element list of a class body would be handled exactly analogous to statement lists where list elements all end with a ; except for those that always end with a }.  However, all of the class body elements we defined happen to end with }.  The empty _ClassElement_ production _ClassElement_ : ; exists as an analog to the _EmptyStatement_ and allows semi-colons to be placed after } terminated elements such as method definitions. ASI automatically applied to class bodies because it is automatically applied everywhere except the header of a `for` statement. Note that the other alternative considered was for the class body to be a comma separated list analogues to the body of an object literal.  The committees preference was to go with the statement list analog.

>Between the three of ... what would be your preference?

There actually could have been a fourth alternative.  The class field ASI hazard is worsen by the lack of a keyword that marks the start of a class field definition. This is why class field definitions are at least as much of a ASI hazard as ExpressStatement in statement bodies.  It's one of the reasons I was opposed to keyword free field definitions.